packages/graphql-codegen-cli/src/utils/patterns.ts

```diff
@@ -0,0 +1,308 @@
+  return patternSets.flatMap(patternSet => [
+    ...patternSet.watch.patterns,
+    ...patternSet.documents.patterns,
+    ...patternSet.schemas.patterns,
+  ]);
+};
+
+/**
+ * Create a rebuild trigger that follows the algorithm described here:
+ * https://github.com/dotansimha/graphql-code-generator/issues/9270#issuecomment-1496765045
+ *
+ * There is a flow chart diagram in that comment.
+ *
+ * Basically:
+ *
+ *  * "Global" patterns are defined at top level of config file, and "local"
+ *    patterns are defined for each output target
+ *  * Each pattern can have "watch", "documents", and "schemas"
+ *  * Watch patterns (global and local) always take precedence over documents and
+ *    schemas patterns, i.e. a watch negation always negates, and a watch match is
+ *    a match even if it would be negated by some pattern in documents or schemas
+ *  * The trigger returns true if any output target's local patterns result in
+ *    a match, after considering the precedence of any global and local negations
+ */
+export const makeShouldRebuild = ({
+  globalPatternSet,
+  localPatternSets,
+}: {
+  globalPatternSet: PatternSet;
+  localPatternSets: PatternSet[];
+}) => {
+  const localMatchers = localPatternSets.map(localPatternSet => {
+    return (path: string) => {
+      // Is path negated by any negating watch pattern?
+      if (
+        matchesAnyAffirmativePattern(
+          path,
+          invertNegatedPatterns([...globalPatternSet.watch.negated, ...localPatternSet.watch.negated])
+        )
+      ) {
+        // Short circut: negations in watch patterns take priority
+        return false;
+      }
+
+      // Does path match any affirmative watch pattern?
+      if (
+        matchesAnyAffirmativePattern(path, [
+          ...globalPatternSet.watch.affirmative,
+          ...localPatternSet.watch.affirmative,
+        ])
+      ) {
+        // Immediately return true: Watch pattern takes priority, even if documents or schema would negate it
+        return true;
+      }
+
+      // Does path match documents patterns (without being negated)?
+      if (pathMatches(path, [...globalPatternSet.documents.patterns, ...localPatternSet.documents.patterns])) {
+        return true;
+      }
+
+      // Does path match schemas patterns (without being negated)?
+      if (pathMatches(path, [...globalPatternSet.schemas.patterns, ...localPatternSet.schemas.patterns])) {
+        return true;
+      }
+
+      // Otherwise, there is no match
+      return false;
+    };
+  });
+
+  /**
+   * Return `true` if `path` should trigger a rebuild
+   */
+  return ({ path: absolutePath }: { path: string }) => {
+    if (!isAbsolute(absolutePath)) {
+      throw new Error('shouldRebuild trigger should be called with absolute path');
+    }
+
+    const path = relative(process.cwd(), absolutePath);
+    const shouldRebuild = localMatchers.some(matcher => matcher(path));
+    return shouldRebuild;
+  };
```

Comment:
This is the callback that is called on each path change. It loops over the pre-constructed matchers, one for each "local" (per-output target) pattern set. This way we avoid too much overhead on every file change, as the functions have all already been created. That's also why we separate the patterns into affirmative/negated only once, when we call `sortPatterns`, and store them in a `PatternSet` struct, rather than separating them on each path change.

Reply:
See https://github.com/dotansimha/graphql-code-generator/pull/9275#discussion_r1181497873